core/http-auth-aws/src/main/java/software/amazon/awssdk/http/auth/aws/chunkedencoding/ChunkedEncodedInputStream.java

```diff
@@ -0,0 +1,303 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.http.auth.aws.chunkedencoding;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import software.amazon.awssdk.annotations.SdkProtectedApi;
+import software.amazon.awssdk.http.auth.aws.internal.chunkedencoding.Chunk;
+import software.amazon.awssdk.utils.Logger;
+import software.amazon.awssdk.utils.Pair;
+import software.amazon.awssdk.utils.Validate;
+
+
+/**
+ * An implementation of chunk-transfer encoding, but by wrapping an {@link InputStream}.
+ * This implementation supports chunk-headers, chunk-extensions, and trailers.
+ * <p>
+ * Per <a href="https://datatracker.ietf.org/doc/html/rfc7230#section-4.1">RFC-7230</a>, a chunk-transfer encoded message
+ * is defined as:
+ * <pre>
+ *     chunked-body   = *chunk
+ *                      last-chunk
+ *                      trailer-part
+ *                      CRLF
+ *     chunk          = chunk-size [ chunk-ext ] CRLF
+ *                      chunk-data CRLF
+ *     chunk-size     = 1*HEXDIG
+ *     last-chunk     = 1*("0") [ chunk-ext ] CRLF
+ *     chunk-data     = 1*OCTET ; a sequence of chunk-size octets
+ * </pre>
+ */
```

Comment:
It feels like this class is doing two things:
1. Mapping an unchunked stream to an chunked stream
2. Allowing setting of the chunk size, extensions and trailers during that transformation.

It feels like it would be better supported with two classes:
1. One to take any InputStream and convert it to a chunked stream
2. One to take a chunked-encoded input stream and transform the individual chunks

Without doing this split, we'd probably still need a separate class for taking a chunked stream and transforming the individual chunks.

Reply:
It's doing (mostly) what the chunked-encoding spec says. Transformation of the body isn't part of chunk transfer encoding.

Reply:
So we'd need a separate class to do any transformation?

Reply:
Yeah, I would say so. Transforming the body *within* this implementation *could* be troublesome. If the body is transformed, then the header/extensions/trailers need some kind of indication of whether they should use the original data or the transformed data. For example, if a chunk is transformed, but the header uses the original to calculate a size, then it may be incorrect. Or if there are multiple transformations, which to use when writing the header/extension/trailers - sounds complex.